crates/iota-data-ingestion-core/src/reader/v2.rs

```diff
@@ -0,0 +1,565 @@
+/// interacting with the running actor, such as receiving checkpoints, sending
+/// GC signals, or triggering shutdown.
+///
+/// # Responsibilities
+/// - Periodically scans for new checkpoints from configured sources.
+/// - Streams checkpoints to consumers via channels.
+/// - Handles garbage collection signals to prune processed checkpoints.
+/// - Coordinates with remote fetchers for batch downloads and retries.
+///
+/// # Usage
+/// Users should not construct or manage `CheckpointReader` directly. Instead,
+/// use [`CheckpointReader::new`] to spawn the actor and obtain a handle
+/// for interaction.
+struct CheckpointReaderActor {
+    /// Filesystem path to the local checkpoint directory.
+    path: PathBuf,
+    /// Start fetch from the current checkpoint sequence.
+    current_checkpoint_number: CheckpointSequenceNumber,
+    /// Keeps tracks the last processed checkpoint sequence number, used to
+    /// delete checkpoint files from ingestion path.
+    last_pruned_watermark: CheckpointSequenceNumber,
+    /// Channel for sending checkpoints to WorkerPools.
+    checkpoint_tx: mpsc::Sender<Arc<CheckpointData>>,
+    /// Sends a garbage collection (GC) signal to prune checkpoint files below
+    /// the specified watermark.
+    gc_signal_rx: mpsc::Receiver<CheckpointSequenceNumber>,
+    /// Remote checkpoint reader for fetching checkpoints from the network.
+    remote_store: Option<Arc<RemoteStore>>,
+    /// Signal when the reader should exit.
+    shutdown_rx: oneshot::Receiver<()>,
+    /// Configures the behavior of the checkpoint reader.
+    reader_options: ReaderOptions,
+    /// Limit the amount of downloaded checkpoints held in memory to avoid OOM.
+    data_limiter: DataLimiter,
+}
+
+impl LocalRead for CheckpointReaderActor {
+    fn exceeds_capacity(&self, checkpoint_number: CheckpointSequenceNumber) -> bool {
+        ((MAX_CHECKPOINTS_IN_PROGRESS as u64 + self.last_pruned_watermark) <= checkpoint_number)
+            || self.data_limiter.exceeds()
+    }
+
+    fn path(&self) -> &Path {
+        &self.path
+    }
+
+    fn current_checkpoint_number(&self) -> CheckpointSequenceNumber {
+        self.current_checkpoint_number
+    }
+
+    fn update_last_pruned_watermark(&mut self, watermark: CheckpointSequenceNumber) {
+        self.last_pruned_watermark = watermark;
+    }
+}
+
+impl CheckpointReaderActor {
+    fn should_fetch_from_remote(&self, checkpoints: &[Arc<CheckpointData>]) -> bool {
+        self.remote_store.is_some()
+            && (checkpoints.is_empty()
+                || checkpoints[0].checkpoint_summary.sequence_number
+                    > self.current_checkpoint_number)
+    }
+
+    /// Fetch a single checkpoint from the live object store.
+    async fn fetch_from_live_object_store(
+        live: &dyn ObjectStore,
+        checkpoint_number: CheckpointSequenceNumber,
+    ) -> CheckpointResult {
+        let path = ObjectStorePath::from(format!("ingestion/live/{checkpoint_number}.chk"));
```

Comment:
Here we should assume that the `ObjectStore` url is the path to the store of `*.chk` files directly. 

Making assumptions on the path being the parent of a `ingestion/live` is rather inflexible for users and error prone.

```suggestion
        let path = ObjectStorePath::from(format!("{checkpoint_number}.chk"));
```

Reply:
I've opted to remove this code entirely in favor of using the existing `reader::fetch::fetch_from_object_store` function. After refactoring, this would have been functionally equivalent, and reusing the existing function eliminates code duplication while also addressing the path flexibility concern you raised.

fixed: https://github.com/iotaledger/iota/pull/7425/commits/5241e16fc87e6e1997c7325af2a426a5a988324d